features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,404 @@
+
+    // compute eigenvalues and eigenvectors of the covariance matrix
+    pcl::eigen33 (covariance_matrix, eigenvectors, eigenvalues);
+
+    // z axis of the reference frame is the eigenvector associated with the minimal eigenvalue
+    Eigen::Vector3f z_axis = eigenvectors.col (0);
+
+    // if angle between z axis and viewing direction is in the [-90 deg, 90 deg] range, then z axis is negated
+    if (z_axis.dot (view_direction_) > 0)
+    {
+      z_axis = z_axis * -1;
+    }
+
+    // x axis of the reference frame is the eigenvector associated with the maximal eigenvalue 
+    Eigen::Vector3f x_axis = eigenvectors.col (2);
+
+    // y axis is the cross product of z axis and x axis
+    Eigen::Vector3f y_axis = z_axis.cross (x_axis);
+
+    Eigen::Vector3f centroid_xyz = centroid.head<3> ();
+
+    // compute alignment transform from axes and centroid
+    transform_ << x_axis[0], x_axis[1], x_axis[2], -x_axis.dot (centroid_xyz), y_axis[0], y_axis[1], y_axis[2], -y_axis.dot (centroid_xyz), z_axis[0], z_axis[1], z_axis[2], -z_axis.dot (
+        centroid_xyz), 0.0f, 0.0f, 0.0f, 1.0f;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::addSampleToHistograms (Eigen::Vector4f &p,
+                                                              float max_coord,
+                                                              int half_grid_size,
+                                                              HistogramInterpolationMethod interp,
+                                                              float hbin,
+                                                              float hist_incr,
+                                                              vector<Eigen::VectorXf> &hists)
+  {
+    int grid_size = half_grid_size * 2;
+
+    // compute normalized coordinates with respect to axis-aligned bounding cube centered on the origin
+    float x_scaled = (p[0] / max_coord) * half_grid_size;
+    float y_scaled = (p[1] / max_coord) * half_grid_size;
+    float z_scaled = (p[2] / max_coord) * half_grid_size;
+
+    // compute histograms array coords
+    float xbin = x_scaled + half_grid_size;
+    float ybin = y_scaled + half_grid_size;
+    float zbin = z_scaled + half_grid_size;
+
+    // if using histogram interpolation, subtract 0.5 so samples with the central value of the bin have full weight in it
+    if (interp != INTERP_NONE)
+    {
+      xbin -= 0.5f;
+      ybin -= 0.5f;
+      zbin -= 0.5f;
+      hbin -= 0.5f;
+    }
+
+    // compute histograms bins indices
+    int x0 = floor (xbin);
+    int y0 = floor (ybin);
+    int z0 = floor (zbin);
+    int h0 = floor (hbin);
+
+    float v_xyz111, v_xyz110, v_xyz101, v_xyz100, v_xyz011, v_xyz010, v_xyz001, v_xyz000;
+
+    // if using histogram interpolation, compute trilinear interpolation
+    if (interp != INTERP_NONE)
```

Comment:
If the comment above is in place this if guard is no longer needed and v_xyz111, v_xyz110, v_xyz101, v_xyz100, v_xyz011, v_xyz010, v_xyz001, v_xyz000 can be initialized as const